server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java

```diff
@@ -617,13 +561,12 @@ private boolean checkForChangedDerivedProducts(Pool pool, Pool existingPool,
             if (pool.getDerivedProduct() != null) {
                 existingPool.setDerivedProduct(pool.getDerivedProduct());
             }
-            else {
-                // subscription no longer has a derived product
+        }
+        else {
+            // check if subscription no longer has a derived product
+            if (pool.getDerivedProduct() == null && existingPool.getDerivedProduct() != null) {
```

Comment:
Earlier most of logic is dependent on provided Products which were directly present
on pool & subscription which is now removed.

In this particular code block, there is one scenario in which derived product is removed from incoming subscription, resulting derived product being set as null on pool, which require a explicit check on pool if its derived product is null and existing pool derived product is not null.

This code block was written inside IF code block and only execute where there is any change in derived product (totally new derived product or it's same derived product with its member properties changed reflected via `changedProducts` map ) which is now `False` as the whole derived product is removed.

Reply:
Right, but we already have a logical block for determining the state, and we're already doing half the check done here in that block. We don't want to separate those blocks if we can avoid it:

```
if (pool.getDerivedProduct() != null) {
  // existing code as above
}
else if (existingPool.getDerivedProduct() != null) {
  productsChanged = true;
}
```

The previous code did work based on the null value (implicitly via encapsulation violations), so it *had* to do them there. Since we don't need to worry about that, we can safely just pass through whatever value is returned by getDerivedProduct, so long as our productsChanged variable is in the right state by the time we get there; we don't even need the additional null check, nor the warning comment block, anymore

Reply:
There's also an argument to be made that we're not handling all of the possible transitions here. As it is now, we handle derived -> derived, and derived -> null, but we still NPE on null -> derived. This is a good opportunity to fix that as well.

Reply:
Refactored it. Yes there is a potential NPE around `existingPool.getDerivedProduct()` in equals statement.